### PR TITLE
Add projections

### PR DESCRIPTION
**Pull request by YuriiHaidamaka**
In this PR we have added proto files for projections. 
Created states for all projections and tests for it.

Also was create integrational tests: the `ExlibrisTest` class. In which we test that after one flow the projections are in the right state.

Due to the fact that there is often a situation where we need to make `LocalDate` from `Timestamp`  we have added a utility class `Timestamps` whic provides a method `toLocalDate` for this aim.

We have refactored command and event factories for testing needs. Created the `TestValues` class with all constants for testing. 

Due to the fact that Inventory events do not contain information about the book details , the following enrichment was added: 
```
message BookEnrichment {
    option (enrichment_for) = "javaclasses.exlibris.c.BookBorrowed,"
                                "javaclasses.exlibris.c.ReservationAdded,"
                                "javaclasses.exlibris.c.BookLost,"
                                "javaclasses.exlibris.c.InventoryAppended,"
                                "javaclasses.exlibris.c.BookReturned,"
                                "javaclasses.exlibris.c.ReservationCanceled,"
                                "javaclasses.exlibris.c.ReservationPickUpPeriodExpired,"
                                "javaclasses.exlibris.c.ReservationBecameLoan,"
                                "javaclasses.exlibris.c.LoanBecameOverdue,"
                                "javaclasses.exlibris.c.LoanBecameShouldReturnSoon,"
                                "javaclasses.exlibris.c.LoanPeriodExtended";

    Book book = 1 [(by) = "inventory_id"];
}
```

In this regard, the `BoundedContest` has been changed and the `ExlibrisEnrichments` class that adds enrichment fields to the `EventBus` has been added.

### CODE REVIEW

File: api-java/src/main/java/javaclasses/exlibris/Timestamps.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        final LocalDate date = LocalDate.newBuilder()
+                                        .setDay(calendar.get(Calendar.DAY_OF_MONTH))
+                                        .setMonth(
+                                                MonthOfYear.valueOf(
+                                                        calendar.get(Calendar.MONTH) + 1))
+                                        .setYear(calendar.get(Calendar.YEAR))
+                                        .build();
+        return date;
+    }
+
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/Timestamps.java]: Remove this line

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookDetailsViewProjection.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        final BookDetails bookDetails = event.getDetails();
+        getBuilder().setBookDetails(bookDetails);
+    }
+
+    @Subscribe
+    public void on(BookUpdated event) {
+        final BookDetails bookDetails = event.getBookDetailsChange()
+                                             .getNewBookDetails();
+        getBuilder().setBookDetails(bookDetails);
+    }
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookDetailsViewProjection.java:58]: Should subscribe on BookRemoved&

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookDetailsViewRepository.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Adds the {@link io.spine.server.route.EventRoute EventRoute}s to the repository.
+     */
+    protected void setUpEventRoute() {
+        final EventRouting<BookId> routing = getEventRouting();
+        routing.route(BookAdded.class,
+                      (message, context) -> singleton(message.getBookId()));
+        routing.route(BookUpdated.class,
+                      (message, context) -> singleton(message.getBookId()));
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookDetailsViewRepository.java:53]: add routing for book removed

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookEventLogViewProjection.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        final Timestamp whenEmitted = event.getWhenReported();
+
+        getBuilder().setBookId(bookId)
+                    .setItemId(itemId)
+                    .setUserName(userName)
+                    .setEmail(email)
+                    .setEventType(eventType)
+                    .setWhenEmitted(whenEmitted);
+    }
+
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookEventLogViewProjection.java:287]: Subscribe on BookReadyToPickUp

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                                                           .setItemId(itemId)
+                                                           .setInLibrary(true)
+                                                           .build();
+        getBuilder().setTitle(title)
+                    .setAuthor(authorName)
+                    .addItemState(state);
+    }
+
+    @Subscribe
+    public void on(InventoryDecreased event) {
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: it looks like subscriber of InventoryRemoved event

Comment (YegorUdovchenko) [reply api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: Add proper Inventory decreased subscriber

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                                                   .setWhenTaken(whenTaken)
+                                                   .setWhenDue(whenDue)
+                                                   .setOverdue(false)
+                                                   .build();
+        final InventoryItemState inventoryItemState = InventoryItemState.newBuilder()
+                                                                        .setItemId(itemId)
+                                                                        .setLoanDetails(loanDetails)
+                                                                        .build();
+        final List<InventoryItemState> items = new ArrayList<>(getBuilder().getItemState());
+        final int index = getIndexByInventoryId(items, event.getInventoryItemId());
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: getIndexByInventoryItemId()

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    public void on(BookLost event) {
+        final InventoryItemId itemId = event.getInventoryItemId();
+
+        final List<InventoryItemState> items = new ArrayList<>(getBuilder().getItemState());
+        final int index = getIndexByInventoryId(items, event.getInventoryItemId());
+        final InventoryItemState invItemState = items.get(index);
+        final InventoryItemState newInvItemState = InventoryItemState.newBuilder(invItemState)
+                                                                     .setItemId(itemId)
+                                                                     .setLost(true)
+                                                                     .build();
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: .clearLoanDetails())

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    @Subscribe
+    public void on(BookReturned event) {
+        final InventoryItemId itemId = event.getInventoryItemId();
+
+        final List<InventoryItemState> items = new ArrayList<>(getBuilder().getItemState());
+        final int index = getIndexByInventoryId(items, event.getInventoryItemId());
+        final InventoryItemState invItemState = items.get(index);
+        final InventoryItemState newInvItemState = InventoryItemState.newBuilder(invItemState)
+                                                                     .setItemId(itemId)
+                                                                     .setInLibrary(true)
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: .clearLoanDetails()

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        final Timestamp whenDueTimestamp = event.getNewDueDate();
+        final LocalDate whenDue = toLocalDate(whenDueTimestamp);
+
+        final List<InventoryItemState> items = new ArrayList<>(getBuilder().getItemState());
+        final int index = getIndexByInventoryId(items, event.getInventoryItemId());
+        final InventoryItemState invItemState = items.get(index);
+
+        final LoanDetails loanDetails = LoanDetails.newBuilder(invItemState.getLoanDetails())
+                                                   .setWhenDue(whenDue)
+                                                   .build();
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewProjection.java]: .clearOverdue()

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewRepository.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                      (message, context) -> singleton(message.getInventoryId()));
+        routing.route(BookLost.class,
+                      (message, context) -> singleton(message.getInventoryId()));
+        routing.route(LoanBecameOverdue.class,
+                      (message, context) -> singleton(message.getInventoryId()));
+        routing.route(BookReturned.class,
+                      (message, context) -> singleton(message.getInventoryId()));
+        routing.route(LoanPeriodExtended.class,
+                      (message, context) -> singleton(message.getInventoryId()));
+    }
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookInventoryViewRepository.java:71]: add routing for InventoryRemoved

File: api-java/src/main/java/javaclasses/exlibris/q/admin/BookLoanViewProjection.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                                              .setFamilyName("Petrov")
+                                              .build();
+
+        getBuilder().setBookId(bookId)
+                    .setItemId(inventoryItemId)
+                    .setUserName(userName)
+                    .setEmail(email)
+                    .setWhenTaken(whenTaken)
+                    .setWhenDue(whenDue);
+    }
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/BookLoanViewProjection.java:76]: Add LoanPeriodExtended subscriber

File: api-java/src/main/java/javaclasses/exlibris/q/admin/LostBookViewRepository.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Adds the {@link io.spine.server.route.EventRoute EventRoute}s to the repository.
+     */
+    protected void setUpEventRoute() {
+        final EventRouting<InventoryItemId> routing = getEventRouting();
+        routing.route(BookLost.class,
+                      (message, context) -> singleton(message.getInventoryItemId()));
+        routing.route(InventoryDecreased.class,
+                      (message, context) -> singleton(message.getInventoryItemId()));
+
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/admin/LostBookViewRepository.java]: remove line

File: api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+        final LoanId loanId = event.getLoanId();
+        final BookId bookId = event.getInventoryId()
+                                   .getBookId();
+        final BorrowedBookItemStatus status = BorrowedBookItemStatus.BORROWED;
+        final BookEnrichment enrichment = getEnrichment(BookEnrichment.class, context);
+        final BookDetails bookDetails = enrichment.getBook()
+                                                  .getBookDetails();
+        final Timestamp whenBorrowedTimeStamp = event.getWhenBorrowed();
+        final LocalDate whenBorrowed = toLocalDate(whenBorrowedTimeStamp);
+        final LocalDate dueDate = addDays(whenBorrowed, 14);
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java]: event has due date

File: api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+                                                       .getBookId());
+        final BorrowedBookItem bookItem = items.get(index);
+        final BorrowedBookItem newBookItem = BorrowedBookItem.newBuilder(bookItem)
+                                                             .setStatus(
+                                                                     BorrowedBookItemStatus.BORROWED)
+                                                             .setIsAllowedLoanExtension(false)
+                                                             .setDueDate(newDueDate)
+                                                             .build();
+        getBuilder().setBookItem(index, newBookItem)
+                    .setNumberOfBorrowedBooks(getBuilder().getBookItem()
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java]: number of borrowed books hasn't changed

File: api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private int getNumberOfOverdueBooks(List<BorrowedBookItem> items) {
+        final int count = (int) items.stream()
+                                     .filter(item -> item.getStatus()
+                                                         .equals(BorrowedBookItemStatus.OVERDUE))
+                                     .count();
+        return count;
+    }
+
```
Comment (YegorUdovchenko) [inline api-java/src/main/java/javaclasses/exlibris/q/user/BorrowedBooksListViewProjection.java]: remove line

File: api-java/src/test/java/javaclasses/exlibris/ExlibrisTest.java [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@DisplayName("Exlibris Integration Test")
+public class ExlibrisTest {
+
+    private final ActorRequestFactory requestFactory =
+            TestActorRequestFactory.newInstance(getClass());
+
+    private BoundedContext boundedContext = BoundedContexts.create();
```
Comment (YegorUdovchenko) [inline api-java/src/test/java/javaclasses/exlibris/ExlibrisTest.java]: not needed to initialize here, BeforeEach method already makes it

**Review by YegorUdovchenko — CHANGES_REQUESTED** [review]
@YuriiHaidamaka PTAL

**Comment by YuriiHaidamaka** [comment]
@YegorUdovchenko PTAL

**Review by YegorUdovchenko — APPROVED** [review]

**Pull request merged**